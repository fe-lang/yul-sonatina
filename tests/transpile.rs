use dir_test::{dir_test, Fixture};
use sonatina_ir::ir_writer::ModuleWriter;
use yul_sonatina::compile;

// copied from fe test-utils
/// A macro to assert that a value matches a snapshot.
/// If the snapshot does not exist, it will be created in the same directory as
/// the test file.
#[macro_export]
macro_rules! snap_test {
    ($value:expr, $fixture_path: expr) => {
        snap_test!($value, $fixture_path, None)
    };

    ($value:expr, $fixture_path: expr, $suffix: expr) => {
        let mut settings = insta::Settings::new();
        let fixture_path = ::std::path::Path::new($fixture_path);
        let fixture_dir = fixture_path.parent().unwrap();
        let fixture_name = fixture_path.file_stem().unwrap().to_str().unwrap();

        settings.set_snapshot_path(fixture_dir);
        settings.set_input_file($fixture_path);
        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);

        let suffix: Option<&str> = $suffix;
        let name = if let Some(suffix) = suffix {
            format!("{fixture_name}.{suffix}")
        } else {
            fixture_name.into()
        };
        settings.bind(|| {
            insta::_macro_support::assert_snapshot(
                (name, $value.as_str()).into(),
                std::path::Path::new(env!("CARGO_MANIFEST_DIR")),
                fixture_name,
                module_path!(),
                file!(),
                line!(),
                &$value,
            )
            .unwrap()
        })
    };
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files",
    glob: "*.yul",
)]
fn test(fixture: Fixture<&str>) {
    compile(fixture.content()).unwrap();

    let module = compile(fixture.content()).unwrap();

    let mut module_writer = ModuleWriter::new(&module);
    let ir = module_writer.dump_string();
    snap_test!(ir, fixture.path());
}
