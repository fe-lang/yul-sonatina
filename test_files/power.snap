---
source: tests/transpile.rs
input_file: test_files/power.yul
---
target = evm-ethereum-cancun

func private %block0::power(v0.i256, v1.i256) -> i256 {
    block0:
        br_table v1 block3 (0.i256 block1) (1.i256 block2);

    block1:
        jump block4;

    block2:
        jump block4;

    block3:
        v6.i256 = mul v0 v0;
        v9.i256 = evm_udiv v1 2.i256;
        v10.i256 = call %block0::power v6 v9;
        v11.i256 = evm_umod v1 2.i256;
        br_table v11 block6 (1.i256 block5);

    block4:
        v15.i256 = phi (1.i256 block1) (v0 block2) (v16 block6);
        return v15;

    block5:
        v14.i256 = mul v0 v10;
        jump block6;

    block6:
        v16.i256 = phi (v10 block3) (v14 block5);
        jump block4;
}

func private %block0::power2(v0.i256, v1.i256) -> i256 {
    block0:
        jump block1;

    block1:
        v11.i256 = phi (v0 block0) (v11 block3);
        v10.i256 = phi (1.i256 block0) (v9 block3);
        v5.i256 = phi (v1 block0) (v5 block3);
        v4.i256 = phi (0.i256 block0) (v13 block3);
        v6.i256 = lt v4 v5;
        br v6 block2 block4;

    block2:
        v9.i256 = mul v10 v11;
        jump block3;

    block3:
        v13.i256 = add v4 1.i256;
        jump block1;

    block4:
        return v10;
}
